src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs

```diff
@@ -46,9 +47,10 @@ public ConsoleExporter(ConsoleExporterOptions options)
             this.serializerOptions.Converters.Add(new ActivityTraceIdConverter());
         }
 
-        public override Task<ExportResult> ExportAsync(IEnumerable<Activity> activityBatch, CancellationToken cancellationToken)
+        [MethodImpl(MethodImplOptions.Synchronized)]
```

Comment:
I didn't even notice this until I read the description. My initial reaction... feels a little hackish. We shouldn't rely on users to add this, we should wrap the call in a lock if we want to make sure it isn't called concurrently? The spec says something about Export shouldn't be called concurrently, it doesn't say Exporters should implement Export so it isn't callable concurrently, which is what this feels like. 🤷

Reply:
Or how about having the `SimpleExportActivityProcessor` taking a parameter which allows people to specify whether it is reentrant or not (and by default use lock), I was thinking about ETW/LTTng scenario where folks want ultimate concurrency.

Something like [this](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.usegloballock).

Reply:
I like the idea of a flag to opt into different behavior than what the spec says. We could also have a SimpleProcessor and a BatchingProcessor that are fully spec-compliant but also ship something like a ConcurrentProcessor or ReentrantProcessor that can be explicit in its changing of the rules?

Reply:
That's works, too! 😄

Reply:
I moved the current logic to `ReentrantExportActivityProcessor`, and changed `SimpleExportActivityProcessor` to simply call the reentrant one with a lock guard.

Reply:
Now the PR got merged and OOP fans are staring at us - "Why are you guys making a non-reentrant class a child of reentrant one" 😂

Reply:
It is a bit odd, isn't it? I know you were shooting for code re-use, maybe the flag/option idea would be better? You could make a base class that has everything but the OnEnd part?

Reply:
Tracked by #898.